db/migration/migration.go

```diff
@@ -0,0 +1,209 @@
+package migration
+
+import (
+	"crypto/rand"
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	bolt "github.com/coreos/bbolt"
+)
+
+const dbFileMode = 0600
+
+type Migration interface {
+	Migrate(db *bolt.DB) error
+	Version() uint32
+}
+
+func randomTempDBPath() (string, error) {
+	file := make([]byte, 50)
+	_, err := rand.Read(file)
+	if err != nil {
+		return "", err
+	}
+	return filepath.Abs(filepath.Join(os.TempDir(), hex.EncodeToString(file)+".bolt"))
+}
+
+func findNextMigration(currentVersion uint32, migrations []Migration) (Migration, error) {
+	mig := doubleMigration(migrations)
+	if mig != nil {
+		return nil, errors.New(fmt.Sprintf("migration %d is already registered", mig.Version()))
+	}
+	// sort the migrations based on there version
+	sort.SliceStable(migrations, func(i, j int) bool {
+		return migrations[i].Version() < migrations[j].Version()
+	})
+	for _, m := range migrations {
+		if currentVersion < m.Version() {
+			return m, nil
+		}
+	}
+	return nil, nil
+}
+
+// returns a migration if a version with the same ID already exist
+func doubleMigration(migrations []Migration) Migration {
+	existingMigrations := map[uint32]bool{}
+	for _, m := range migrations {
+		_, exist := existingMigrations[m.Version()]
+		if exist {
+			return m
+		}
+		existingMigrations[m.Version()] = true
+	}
+	return nil
+}
+
+// migrate migrates a bold db database
+func Migrate(prodDBPath string, migrations []Migration) error {
+
+	// make sure there are no double migrations
+	mig := doubleMigration(migrations)
+	if mig != nil {
+		return errors.New(fmt.Sprintf("a migration with id (%d) was already registered", mig.Version()))
+	}
+
+	// check if production database exist
+	if _, err := os.Stat(prodDBPath); err != nil {
+		return err
+	}
+
+	// open production database
+	prodDB, err := bolt.Open(prodDBPath, dbFileMode, &bolt.Options{Timeout: time.Second})
```

Comment:
Since &bolt.Options{Timeout: time.Second} is essentially &bolt.Options{Timeout: 1 * time.Second}, maybe we should use the latter as it demonstrates that the timeout have intentionally been set to 1 second.

Reply:
added